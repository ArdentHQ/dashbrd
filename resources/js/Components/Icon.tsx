import React, { type ElementType } from "react";

import { SvgCollection } from "@/icons";

type Size = "xs" | "sm" | "md" | "lg" | "xl" | "2xl";

export type IconName = keyof typeof SvgCollection;

interface IconProperties extends React.HTMLAttributes<SVGElement> {
    name: IconName;
    size?: Size;
    as?: ElementType;
    fallback?: JSX.Element;
    dimensions?: IconDimensions;
    className?: string;
}

export interface IconDimensions {
    width: number;
    height: number;
}

const getDimensions = (size?: Size, dimensions?: IconDimensions): IconDimensions => {
    if (dimensions !== undefined) {
        return dimensions;
    }

    const sizeMap: Record<string, IconDimensions> = {
        "2xl": {
            width: 40,
            height: 40,
        },
        xl: {
            width: 32,
            height: 32,
        },
        lg: {
            width: 20,
            height: 20,
        },
        md: {
            width: 16,
            height: 16,
        },
        sm: {
            width: 12,
            height: 12,
        },
        xs: {
            width: 8,
            height: 8,
        },
    };

    return sizeMap[size ?? "md"];
};

export const Icon = ({ name, size, dimensions, ...properties }: IconProperties): JSX.Element => {
    const Svg = SvgCollection[name];

    dimensions = getDimensions(size, dimensions);

    return (
        <Svg
            data-testid={`icon-${name}`}
            width={dimensions.width}
            height={dimensions.height}
            {...properties}
        />
    );
};
