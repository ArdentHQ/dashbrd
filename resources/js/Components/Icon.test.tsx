import React from "react";
import { Icon } from "./Icon";
import { render } from "@/Tests/testing-library";

describe("Icon", () => {
    it("should render", () => {
        const { asFragment } = render(<Icon name="ArrowUp" />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render with dimensions", () => {
        const { asFragment } = render(
            <Icon
                name="ArrowUp"
                dimensions={{ width: 10, height: 10 }}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render with dimensions and default size", () => {
        const { asFragment } = render(
            <Icon
                name="ArrowUp"
                size={undefined}
                dimensions={{ width: 10, height: 10 }}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
