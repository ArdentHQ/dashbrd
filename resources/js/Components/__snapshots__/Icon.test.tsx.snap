// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon > should render 1`] = `
<DocumentFragment>
  <svg
    data-testid="icon-ArrowUp"
    fill="none"
    height="16"
    id="arrow-up"
    viewBox="0 0 8 12"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.028 11V1.6m3.107 3.108L4.028 1.6.92 4.708"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.6"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon > should render with dimensions 1`] = `
<DocumentFragment>
  <svg
    data-testid="icon-ArrowUp"
    fill="none"
    height="10"
    id="arrow-up"
    viewBox="0 0 8 12"
    width="10"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.028 11V1.6m3.107 3.108L4.028 1.6.92 4.708"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.6"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon > should render with dimensions and default size 1`] = `
<DocumentFragment>
  <svg
    data-testid="icon-ArrowUp"
    fill="none"
    height="10"
    id="arrow-up"
    viewBox="0 0 8 12"
    width="10"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.028 11V1.6m3.107 3.108L4.028 1.6.92 4.708"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.6"
    />
  </svg>
</DocumentFragment>
`;
